server/lib/flow-event.js

```diff
@@ -23,7 +24,98 @@ var HOSTNAME = os.hostname();
 var MAX_DATA_LENGTH = 100;
 var VERSION = 1;
 
-module.exports = function (event, data, request) {
+const FLOW_BEGIN_EVENT_TYPES = /^flow\.[a-z_-]+\.begin$/;
+const FLOW_ID_KEY = config.get('flow_id_key');
+const FLOW_ID_EXPIRY = config.get('flow_id_expiry');
+
+const ENTRYPOINT_PATTERN = /^[\w\.-]+$/;
+const SERVICE_PATTERN = /^(sync|content-server|none|[0-9a-f]{16})$/;
+const VALID_FLOW_EVENT_PROPERTIES = [
+  { key: 'client_id', pattern: SERVICE_PATTERN },
+  { key: 'context', pattern: /^[0-9a-z_-]+$/ },
+  { key: 'entryPoint', pattern: ENTRYPOINT_PATTERN },
+  { key: 'entrypoint', pattern: ENTRYPOINT_PATTERN },
+  { key: 'flowId', pattern: /^[0-9a-f]{64}$/ },
+  { key: 'migration', pattern: /^(sync11|amo|none)$/ },
+  { key: 'service', pattern: SERVICE_PATTERN }
+];
+
+const IS_DISABLED = config.get('client_metrics').stderr_collector_disabled;
+
+module.exports = (req, metrics, requestReceivedTime) => {
+  if (IS_DISABLED || ! isValidFlowData(metrics, requestReceivedTime)) {
+    return;
+  }
+
+  const events = metrics.events || [];
+  const flowEvents = _.filter(events, event => {
+    return event.type.indexOf('flow.') === 0;
+  });
+
+  flowEvents.forEach(event => {
+    if (FLOW_BEGIN_EVENT_TYPES.test(event.type)) {
+      event.time = metrics.flowBeginTime;
+      event.flowTime = 0;
+    } else {
+      event.time = estimateTime({
+        /*eslint-disable sorting/sort-object-props*/
+        start: metrics.startTime,
+        offset: event.offset,
+        sent: metrics.flushTime,
+        received: requestReceivedTime
+        /*eslint-enable sorting/sort-object-props*/
+      });
+
+      if (! isValidTime(event.time, requestReceivedTime)) {
+        return;
+      }
+
+      event.flowTime = event.time - metrics.flowBeginTime;
+    }
+
+    logFlowEvent(event, metrics, req);
+  });
+};
+
+function isValidFlowData (metrics, requestReceivedTime) {
+  if (! metrics.flowId) {
+    return false;
+  }
+
+  if (! isValidTime(parseInt(metrics.flowBeginTime), requestReceivedTime)) {
+    return false;
+  }
+
+  if (! VALID_FLOW_EVENT_PROPERTIES.every(p => isValidProperty(metrics[p.key], p.pattern))) {
```

Comment:
They get eliminated by [this call to `_.pick`](https://github.com/mozilla/fxa-content-server/pull/4411/files#diff-426670954645c673de319a60a241f7c4R119). I'm wondering if that would be clearer by validating *after* `pick` instead of before.